# Data Quality Assessment

This notebook conducts a structured data quality assessment of the NovaCred credit application dataset used in automated loan decision-making. The objective is to evaluate whether the dataset provides a reliable and consistent foundation for:

- Credit approval decisions
- Downstream modelling and analytics
- Transparent and explainable decision processes
- Internal governance and quality control

This notebook represents the near-final analytical report for the **data quality pillar** of our project. While the README consolidates overall conclusions across all governance dimensions, this document provides the detailed technical analysis, structured findings, and risk assessment underlying those conclusions.


# Table of Contents

1. [Executive Summary](#1-executive-summary)
2. [Assessment Framework](#2-assessment-framework)
3. [Methodological Principles](#3-methodological-principles)
4. [Dataset Overview](#4-dataset-overview)
5. [Completeness Assessment](#5-completeness-assessment)
6. [Consistency Assessment](#6-consistency-assessment)
7. [Validity Assessment](#7-validity-assessment)
8. [Accuracy and Plausibility Assessment](#8-accuracy-and-plausibility-assessment)
9. [Consolidated Risk Assessment](#9-consolidated-risk-assessment)
10. [Recommended Controls and Remediation Plan](#10-recommended-controls-and-remediation-plan)


# 1. Executive Summary

tbd

# 2. Assessment Framework

The assessment is structured along four established data quality dimensions:


1. **Completeness:** Whether required attributes are present and populated, particularly those relevant for creditworthiness evaluation and decision integrity
2. **Consistency:** Internal coherence across fields, including duplicate identifiers, contradictory values, and structurally inconsistent records
3. **Validity:** Conformity of values to expected formats, ranges, and domain constraints (e.g., financial ratios, identifiers, categorical encodings)
4. **Accuracy:** Whether values appear realistic when assessed in relation to other variables and domain expectations


For each identified issue, we report:

| Attribute | Description |
|---|---|
| Affected records | Count and percentage of dataset |
| Severity | Low · Moderate · High · Critical |
| Governance impact | Effect on decision integrity, auditability, or model reliability |
| Remediation | Recommended corrective measures |


For classifying the severiy of the issues identified, we use the following **severity classification scale:**

| Level | Definition |
|---|---|
| **Low** | Minor data hygiene issue with limited practical impact |
| **Moderate** | Relevant issue requiring correction but unlikely to materially affect decisions |
| **High** | Issue likely to affect decision integrity, model reliability, or auditability |
| **Critical** | Strong indication of control failure with material impact on decision processes |

# 3. Methodological Principles

| Principle | Description |
|---|---|
| **Reproducibility** | All calculations use fixed reference points where applicable |
| **Quantification** | Findings are expressed in both absolute counts and percentages |
| **Materiality** | Issues are evaluated not only statistically but in terms of practical impact on decision-making |
| **Structured reporting** | Each assessment section concludes with a standardized risk assessment table summarizing findings and proposed controls |

# to be deleted later:

**Required Structure for Each Chapter**

For each audit section (Completeness, Consistency, Validity, Accuracy), report findings using the structure below.

**1. Analytical Objective**
State what is being checked and why it matters for governance and decision integrity.

**2. Method and Criteria**
Describe:
- The rule(s) used (e.g., missingness threshold, format constraints, range checks)
- Any benchmarks or thresholds used for risk classification

**3. Results**
Report:
- Key counts and percentages
- A concise table of the most important issues

**4. Interpretation**
Explain:
- Why the issues matter (impact on decision-making, auditability, compliance)
- Whether the issues are structural or anomalous

**5. Risk Assessment**
Include the following table after each chapter:

| Finding | Evidence (Metric) | Threshold / Expectation | Severity | Governance Impact | Recommended Control |
|--------|--------------------|--------------------------|----------|-------------------|---------------------|
| Example: Missing income values | 12 records (2.4%) | Income required for creditworthiness | High | Risk of unreliable decisions and improper approvals | Block decision if income missing; implement validation and monitoring |

---

**Consolidated Risk Assessment (Final Summary)**

At the end of the notebook, consolidate the chapter-level findings into a final summary table:

| Dimension | Issues Identified | Percentage of Records Affected | Highest Severity | Overall Risk |
|----------|-------------------|-------------------------------|-----------------|--------------|
| Completeness | TBD | TBD | TBD | TBD |
| Consistency | TBD | TBD | TBD | TBD |
| Validity | TBD | TBD | TBD | TBD |
| Accuracy | TBD | TBD | TBD | TBD |

Then provide a short written overall assessment:

- Overall data quality risk level: Low / Moderate / Elevated / High
- Dataset suitability for governance and modelling: Acceptable / Conditional / Not acceptable without remediation
- Immediate remediation required: Yes / No

# 4. Dataset Overview

In this section, we provide a short overview of the dataset before conducting the detailed quality assessments. The objective is to establish baseline context regarding dataset size, structure, variable types, and the distribution of key decision outcomes.
This overview serves as a foundation for the subsequent completeness, consistency, validity, and plausibility assessments and ensures that later findings are interpreted in the correct structural context.

In [40]:
# importing necessary libraries and modules
import sys
import os
import pandas as pd
import numpy as np
from src.data_loading import load_raw_data
sys.path.append(os.path.abspath(".."))

In [4]:
# loading the raw data and checking its shape
df = load_raw_data("../data/raw/raw_credit_applications.json")
df.shape

(502, 34)

In [8]:
# checking the columns, data types and non-null counts of the dataframe
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 502 entries, 0 to 501
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   id                            502 non-null    str    
 1   processing_timestamp          62 non-null     str    
 2   loan_purpose                  50 non-null     str    
 3   notes                         2 non-null      str    
 4   full_name                     502 non-null    str    
 5   email                         502 non-null    str    
 6   ssn                           497 non-null    str    
 7   ip_address                    497 non-null    str    
 8   gender                        501 non-null    str    
 9   date_of_birth                 501 non-null    str    
 10  zip_code                      501 non-null    str    
 11  annual_income                 497 non-null    object 
 12  credit_history_months         502 non-null    int64  
 13  debt_to_income  

In [10]:
# checking the summary statistics of the dataframe
df.describe()

,credit_history_months,debt_to_income,savings_balance,spending_shopping,spending_rent,spending_alcohol,interest_rate,approved_amount,spending_dining,spending_healthcare,spending_fitness,spending_entertainment,spending_insurance,spending_travel,spending_transportation,spending_utilities,spending_groceries,spending_education,spending_adult_entertainment,spending_gambling
count,502.000000,502.000000,502.000000,54.00000,59.000000,11.000000,292.000000,292.000000,66.000000,68.000000,71.000000,72.000000,68.000000,80.000000,61.000000,76.000000,65.000000,64.000000,5.000000,7.000000
mean,50.402390,0.246195,29493.503984,456.37037,523.559322,477.090909,4.564726,47845.890411,472.787879,451.367647,455.380282,496.013889,473.176471,493.400000,439.032787,498.605263,487.815385,499.656250,591.200000,457.142857
std,31.234824,0.136296,16775.309756,234.72853,236.866634,221.196046,1.162866,18103.754530,237.168198,250.460049,233.185908,240.582479,230.782442,250.949014,241.350435,246.217036,242.465841,255.760008,200.966913,242.843865
min,-10.000000,0.050000,-5000.000000,73.00000,81.000000,194.000000,2.500000,15000.000000,75.000000,51.000000,61.000000,52.000000,54.000000,54.000000,52.000000,62.000000,112.000000,54.000000,322.000000,125.000000
25%,27.250000,0.150000,17258.250000,269.50000,323.500000,247.000000,3.500000,34000.000000,266.000000,232.500000,253.500000,291.250000,310.750000,255.000000,288.000000,295.750000,254.000000,268.750000,497.000000,277.500000
50%,48.000000,0.230000,27385.500000,443.00000,526.000000,447.000000,4.550000,48000.000000,462.000000,464.000000,444.000000,489.000000,503.000000,532.000000,385.000000,494.500000,503.000000,528.000000,579.000000,444.000000
75%,72.000000,0.350000,38251.500000,648.50000,705.500000,672.000000,5.600000,62250.000000,650.250000,603.500000,614.500000,715.750000,671.500000,703.500000,663.000000,729.750000,691.000000,713.750000,710.000000,662.500000
max,133.000000,1.850000,88078.000000,887.00000,897.000000,757.000000,6.500000,80000.000000,897.000000,897.000000,894.000000,883.000000,890.000000,888.000000,899.000000,881.000000,885.000000,889.000000,848.000000,751.000000


In [15]:
# checking the distribution of the target variable 'loan_approved'
print(df.loan_approved.value_counts())

loan_approved
True     292
False    210
Name: count, dtype: int64


**Findings:**

Dataset Size and Structure
- The dataset contains 502 records and 34 variables
- The structure reflects a flattened representation of nested credit application data, including applicant information, financial attributes, detailed spending categories, and loan decision outcomes

**Data Types and Structural Characteristics**

- The dataset contains a mix of numeric, boolean, and string fields
- Most financial variables are stored as numeric types (float64 or int64)
- At least one financial field (`annual_income`) appears as an object type, which may indicate parsing inconsistencies and will require closer inspection in the validity assessment
- Spending categories show substantial variation in non-null counts, suggesting heterogeneous availability of behavioral data across applicants

**Numerical Distributions**

- `credit_history_months` has a mean of ~50 months but includes a minimum value of -10, which is logically implausible and indicates potential data integrity issues
- `debt_to_income` ranges up to 1.85, suggesting some applicants report debt exceeding income (185%), which may require plausibility evaluation
- `savings_balance` includes negative values (minimum = -5,000), which is unusual and requires validation
- Several spending variables exhibit wide ranges (e.g., values close to 900), suggesting substantial heterogeneity in reported monthly spending behavior

These initial observations indicate potential plausibility and validity concerns that will be systematically examined in later sections.

**Decision Distribution**

The dataset is moderately balanced (Loan approvals: 292 (58.2%); Loan rejections: 210 (41.8%)), with a slight majority of approved applications. This distribution provides a stable baseline for downstream analysis and will serve as a reference point when assessing bias and consistency in decision outcomes.

# 5. Completeness Assessment

In this section, we analyze the completeness of the dataset by assessing the presence of missing values across all relevant fields. The objective is to identify incomplete records that may affect downstream analyses or decision-making.

Completeness is assessed with particular focus on:
- Decision-critical financial variables
- Applicant identity and demographic attributes
- Fields required for traceability and auditability
- Structural integrity of nested components after flattening

We systematically identify missing values, quantify their prevalence, and assess whether the observed missingness poses material risk to decision integrity or governance standards.

In [51]:
missing_overview = (
    df.isna()
      .mean()
      .sort_values(ascending=False)
      .rename("missing_rate")
      .to_frame()
)

missing_overview["missing_pct"] = (missing_overview["missing_rate"] * 100).round(2)
missing_overview["missing_count"] = df.isna().sum()
missing_overview["non_null_count"] = df.notna().sum()
missing_overview = missing_overview[["missing_count", "non_null_count", "missing_pct"]]

print("Global missingness overview:")
display(missing_overview)

Global missingness overview:


,missing_count,non_null_count,missing_pct
notes,500,2,99.60
spending_adult_entertainment,497,5,99.00
spending_gambling,495,7,98.61
spending_alcohol,491,11,97.81
loan_purpose,452,50,90.04
spending_shopping,448,54,89.24
spending_rent,443,59,88.25
spending_transportation,441,61,87.85
processing_timestamp,440,62,87.65
spending_education,438,64,87.25


**Findings:**

The global missingness overview shows that many variables exhibit high missingness rates, with more than half of all columns exceeding 80% missing values. This pattern is primarily driven by the spending_ variables, which individually show 84–99% missingness. Given the variables' nature, a record-level analysis will be performed in the following steps.

Outside the spending variables, loan_purpose stands out with 90% missing values. Given that 58% of applications were approved, this raises interpretability and governance concerns.

To assess materiality appropriately, we therefore group variables into governance-relevant categories for deeper analysis:
1. Decision-Critical Financial Fields
2. Decision Outcome Fields
3. Identity and Traceability Fields
4. Spending Fields

In [52]:
# defining field groups for focused data quality checks
decision_critical_financials = [
    "annual_income",
    "debt_to_income",
    "credit_history_months",
    "savings_balance",
]

decision_outcome_fields = [
    "loan_approved",
    "interest_rate",
    "approved_amount",
    "rejection_reason",
]

identity_traceability_fields = [
    "id",
    "ssn",
    "email",
]

spending_fields = [col for col in df.columns if col.startswith("spending_")]

In [53]:
# printing the defined field groups for reference

print(f"  Decision-critical financials: {decision_critical_financials}")
print(f"  Decision outcome fields:      {decision_outcome_fields}")
print(f"  Identity/traceability fields: {identity_traceability_fields}")
print(f"  Spending fields:              {spending_fields}")

  Decision-critical financials: ['annual_income', 'debt_to_income', 'credit_history_months', 'savings_balance']
  Decision outcome fields:      ['loan_approved', 'interest_rate', 'approved_amount', 'rejection_reason']
  Identity/traceability fields: ['id', 'ssn', 'email']
  Spending fields:              ['spending_shopping', 'spending_rent', 'spending_alcohol', 'spending_dining', 'spending_healthcare', 'spending_fitness', 'spending_entertainment', 'spending_insurance', 'spending_travel', 'spending_transportation', 'spending_utilities', 'spending_groceries', 'spending_education', 'spending_adult_entertainment', 'spending_gambling']


In [54]:
# building a function to check missingness in a given list of fields and print the results
def check_missingness(fields, df=df):
    missing_overview = (
        df[fields]
          .isna()
          .mean()
          .sort_values(ascending=False)
          .rename("missing_rate")
          .to_frame()
    )

    missing_overview["missing_pct"] = (missing_overview["missing_rate"] * 100).round(2)
    missing_overview["missing_count"] = df[fields].isna().sum()
    missing_overview["non_null_count"] = df[fields].notna().sum()
    missing_overview = missing_overview[["missing_count", "non_null_count", "missing_pct"]]

    print(f"Missingness overview for fields: {fields}")
    display(missing_overview)

In [55]:
# checking missingness for the defined field groups
check_missingness(decision_critical_financials)
check_missingness(decision_outcome_fields)
check_missingness(identity_traceability_fields)
check_missingness(spending_fields)

Missingness overview for fields: ['annual_income', 'debt_to_income', 'credit_history_months', 'savings_balance']


,missing_count,non_null_count,missing_pct
annual_income,5,497,1.0
debt_to_income,0,502,0.0
credit_history_months,0,502,0.0
savings_balance,0,502,0.0


Missingness overview for fields: ['loan_approved', 'interest_rate', 'approved_amount', 'rejection_reason']


,missing_count,non_null_count,missing_pct
rejection_reason,292,210,58.17
interest_rate,210,292,41.83
approved_amount,210,292,41.83
loan_approved,0,502,0.00


Missingness overview for fields: ['id', 'ssn', 'email']


,missing_count,non_null_count,missing_pct
ssn,5,497,1.0
id,0,502,0.0
email,0,502,0.0


Missingness overview for fields: ['spending_shopping', 'spending_rent', 'spending_alcohol', 'spending_dining', 'spending_healthcare', 'spending_fitness', 'spending_entertainment', 'spending_insurance', 'spending_travel', 'spending_transportation', 'spending_utilities', 'spending_groceries', 'spending_education', 'spending_adult_entertainment', 'spending_gambling']


,missing_count,non_null_count,missing_pct
spending_adult_entertainment,497,5,99.00
spending_gambling,495,7,98.61
spending_alcohol,491,11,97.81
spending_shopping,448,54,89.24
spending_rent,443,59,88.25
spending_transportation,441,61,87.85
spending_education,438,64,87.25
spending_groceries,437,65,87.05
spending_dining,436,66,86.85
spending_healthcare,434,68,86.45


In [56]:
# checking conditional completeness for decision outcome fields based on the loan approval status

# normalize rejection_reason: treat empty/whitespace as missing
if "rejection_reason" in df.columns:
    df["rejection_reason"] = df["rejection_reason"].replace(r"^\s*$", np.nan, regex=True)

# building denominators
n_total = len(df)
n_approved = int((df["loan_approved"] == True).sum())
n_rejected = int((df["loan_approved"] == False).sum())

# approved loans: required fields must be present
approved_missing_amount = df.query("loan_approved == True and approved_amount.isna()")
approved_missing_interest = df.query("loan_approved == True and interest_rate.isna()")
approved_with_rejection_reason = df.query("loan_approved == True and rejection_reason.notna()")

# rejected loans: rejection_reason must be present; approval fields should be absent
rejected_missing_reason = df.query("loan_approved == False and rejection_reason.isna()")
rejected_with_interest = df.query("loan_approved == False and interest_rate.notna()")
rejected_with_amount = df.query("loan_approved == False and approved_amount.notna()")

# building a compact audit table
checks = [
    ("Approved but approved_amount is missing", len(approved_missing_amount), n_approved),
    ("Approved but interest_rate is missing", len(approved_missing_interest), n_approved),
    ("Approved but rejection_reason is present", len(approved_with_rejection_reason), n_approved),
    ("Rejected but rejection_reason is missing", len(rejected_missing_reason), n_rejected),
    ("Rejected but interest_rate is present", len(rejected_with_interest), n_rejected),
    ("Rejected but approved_amount is present", len(rejected_with_amount), n_rejected),
]

conditional_completeness = pd.DataFrame(checks, columns=["rule", "affected_records", "group_denominator"])

conditional_completeness["affected_pct_within_group"] = (
    conditional_completeness["affected_records"] / conditional_completeness["group_denominator"] * 100
).round(2).replace([np.inf, -np.inf], np.nan)

conditional_completeness["affected_pct_of_all"] = (
    conditional_completeness["affected_records"] / n_total * 100
).round(2)

print(f"Denominators: total={n_total}, approved={n_approved}, rejected={n_rejected}")
display(conditional_completeness)

Denominators: total=502, approved=292, rejected=210


,rule,affected_records,group_denominator,affected_pct_within_group,affected_pct_of_all
0,Approved but approved_amount is missing,0,292,0.0,0.0
1,Approved but interest_rate is missing,0,292,0.0,0.0
2,Approved but rejection_reason is present,0,292,0.0,0.0
3,Rejected but rejection_reason is missing,0,210,0.0,0.0
4,Rejected but interest_rate is present,0,210,0.0,0.0
5,Rejected but approved_amount is present,0,210,0.0,0.0


In [57]:
# analyzing spending data completeness and patterns

if not spending_fields:
    print("No spending_* columns detected. Skipping spending completeness checks.")
else:
    n_total = len(df)

    # Per-row coverage: how many spending categories are populated for each applicant
    spending_non_null_per_row = df[spending_fields].notna().sum(axis=1)
    has_any_spending = spending_non_null_per_row > 0

    spending_coverage = pd.DataFrame({
        "metric": [
            "Records with any spending data",
            "Records with no spending data",
            "Average non-null spending categories per record",
            "Median non-null spending categories per record",
            "Maximum non-null spending categories per record",
        ],
        "value": [
            int(has_any_spending.sum()),
            int((~has_any_spending).sum()),
            round(float(spending_non_null_per_row.mean()), 2),
            float(spending_non_null_per_row.median()),
            int(spending_non_null_per_row.max()),
        ],
        "pct_of_all": [
            round(float(has_any_spending.mean() * 100), 2),
            round(float((~has_any_spending).mean() * 100), 2),
            np.nan,
            np.nan,
            np.nan,
        ]
    })

    print("Spending data coverage summary:")
    display(spending_coverage)

    # Per-category missingness
    spending_missing = pd.DataFrame({
        "missing_count": df[spending_fields].isna().sum(),
        "non_null_count": df[spending_fields].notna().sum(),
        "missing_pct": (df[spending_fields].isna().mean() * 100).round(2),
    }).sort_values("missing_pct", ascending=False)

    spending_missing.index.name = "spending_category"

    print("Missingness per spending category")
    display(spending_missing)

Spending data coverage summary:


,metric,value,pct_of_all
0,Records with any spending data,502.00,100.0
1,Records with no spending data,0.00,0.0
2,Average non-null spending categories per record,1.65,NaN
3,Median non-null spending categories per record,1.00,NaN
4,Maximum non-null spending categories per record,4.00,NaN


Missingness per spending category


,missing_count,non_null_count,missing_pct
spending_category,,,
spending_adult_entertainment,497,5,99.00
spending_gambling,495,7,98.61
spending_alcohol,491,11,97.81
spending_shopping,448,54,89.24
spending_rent,443,59,88.25
spending_transportation,441,61,87.85
spending_education,438,64,87.25
spending_groceries,437,65,87.05
spending_dining,436,66,86.85


**Findings:**

The dataset demonstrates high structural completeness for decision-critical and outcome-related variables. Conditional checks confirm internal completeness logic aligned with approval status.

However:
- `annual_income` is missing for 1% of records. Given its central role in creditworthiness evaluation, this requires explicit governance treatment
- `loan_purpose` and processing_timestamp exhibit high missingness, suggesting structural gaps in data collection processes rather than isolated anomalies
- `Spending_` categories are highly sparse individually, but record-level coverage indicates partial behavioral information is present for all applicants

The absence of violations in conditional decision fields is a strong indicator of structured data capture in the core credit workflow.


**Risk Assessment:**

| Finding | Evidence (Metric) | Threshold / Expectation | Severity | Governance Impact | Recommended Control |
|---|---|---|---|---|---|
| Missing `annual_income` | 5 records (1.0%) | Income required for credit evaluation | **Moderate** | Potential distortion of model training or decision logic | Flag missing income; exclude from modelling or require completion before decision |
| High missingness in `loan_purpose` | 452 records (90.04%) | Purpose typically required for audit traceability | **Moderate** | Reduced interpretability and reduced segmentation capability | Review collection process; clarify whether optional |
| High missingness in `processing_timestamp` | 440 records (87.65%) | Timestamp required for process traceability | **High** | Weak audit trail for processing lifecycle | Implement mandatory timestamp logging |
| Sparse individual spending categories | 84–99% missing per category | Behavioral data optional but informative | **Low** | Limited modeling signal but no direct governance breach | Treat as optional features; avoid imputation |

The dataset demonstrates strong completeness for core credit decision logic, with no conditional violations detected.

Primary governance risks arise from:
- Missing income in a minority of cases
- Structural gaps in metadata (processing timestamps)
- High sparsity in behavioral spending categories

Overall, completeness risk is assessed as **Moderate**, driven primarily by metadata and traceability gaps rather than core financial decision variables.

### 2. Consistency
In this section, internal consistency is evaluated by examining whether records are logically coherent and structurally unique. This includes identifying duplicate application IDs and detecting contradictions between related fields that may compromise data integrity.

In [59]:
# duplicates
df.duplicated().sum()


np.int64(0)

In [60]:
# id based duplicates
df["id"].duplicated().sum()

np.int64(2)

In [61]:
dup_ids = df.loc[df["id"].duplicated(keep=False), ["id", "processing_timestamp", "email", "ssn", "loan_approved", "approved_amount", "interest_rate"]]
dup_ids.sort_values(["id", "processing_timestamp"])

,id,processing_timestamp,email,ssn,loan_approved,approved_amount,interest_rate
383,app_001,None,stephanie.nguyen47@mail.com,427-90-1892,False,NaN,NaN
455,app_001,None,stephanie.nguyen47@mail.com,None,False,NaN,NaN
8,app_042,None,joseph.lopez1@gmail.com,652-70-5530,False,NaN,NaN
354,app_042,None,joseph.lopez1@gmail.com,652-70-5530,False,NaN,NaN


In [62]:
if "gender" in df.columns:
    print("gender")
    print(df["gender"].value_counts(dropna=False))

gender
gender
Male      195
Female    193
F          58
M          53
            2
None        1
Name: count, dtype: int64


In [63]:
'''
The consistency analysis identified 2 duplicate application IDs, while no full row-level duplicates were detected. Although the duplicated records appear largely similar, repeated IDs violate the uniqueness requirement and may compromise data integrity and traceability.
In addition, inconsistent categorical encoding was detected in the gender field. The same logical categories are represented using multiple labels (e.g., “Male” and “M”, “Female” and “F”), indicating a lack of standardized data entry. Such inconsistencies may distort aggregation results and bias downstream analysis.
Remediation should either flag duplicate IDs for review or deduplicate based on a defined business rule (e.g., keeping the first or most recent record). Furthermore, categorical values should be standardized to ensure consistent encoding.

Option A - Flag duplicates:
df["flag_duplicate_id"] = df["id"].duplicated(keep=False)

Option B - Remove duplicates:
df_deduplicated = df.drop_duplicates(subset="id", keep="first")

Standardize gender encoding:
df["gender"] = df["gender"].replace({
    "M": "Male",
    "F": "Female",
    "": pd.NA
})
'''

'\nThe consistency analysis identified 2 duplicate application IDs, while no full row-level duplicates were detected. Although the duplicated records appear largely similar, repeated IDs violate the uniqueness requirement and may compromise data integrity and traceability.\nIn addition, inconsistent categorical encoding was detected in the gender field. The same logical categories are represented using multiple labels (e.g., “Male” and “M”, “Female” and “F”), indicating a lack of standardized data entry. Such inconsistencies may distort aggregation results and bias downstream analysis.\nRemediation should either flag duplicate IDs for review or deduplicate based on a defined business rule (e.g., keeping the first or most recent record). Furthermore, categorical values should be standardized to ensure consistent encoding.\n\nOption A - Flag duplicates:\ndf["flag_duplicate_id"] = df["id"].duplicated(keep=False)\n\nOption B - Remove duplicates:\ndf_deduplicated = df.drop_duplicates(subset

### 3. Validity
In this section, data validity is assessed by verifying whether values comply with predefined formats, ranges, and business rules. This includes checking for invalid formats (e.g., email, ZIP, SSN) and out-of-bound numerical values such as negative income or implausible financial ratios.

In [64]:
import pandas as pd

income = pd.to_numeric(df["annual_income"], errors="coerce")

# negative income
neg_income = (income < 0).sum()

# debt_to_income outside [0,1]
bad_dti = ((df["debt_to_income"] < 0) | (df["debt_to_income"] > 1)).sum()

# negative interest rate
neg_ir = (df["interest_rate"] < 0).sum()

neg_income, bad_dti, neg_ir

(np.int64(0), np.int64(1), np.int64(0))

In [65]:
#check format related validity issues
invalid_income_format = df["annual_income"].notna().sum() - income.notna().sum()
invalid_income_format

np.int64(0)

In [66]:
# negative spending
neg_spending = (df[spend_cols] < 0).sum().sum()

# extreme values (z.B. > 1 Mio)
extreme_spending = (df[spend_cols] > 1_000_000).sum().sum()

neg_spending, extreme_spending

(np.int64(0), np.int64(0))

In [67]:
# Validate that SSN values follow the required XXX-XX-XXXX format
import re

invalid_ssn = df["ssn"].dropna().apply(
    lambda x: not bool(re.fullmatch(r"\d{3}-\d{2}-\d{4}", str(x)))
).sum()

invalid_ssn

np.int64(0)

In [68]:
# Validate that ZIP codes contain only numeric characters
invalid_zip = df["zip_code"].dropna().apply(
    lambda x: not str(x).isdigit()
).sum()

invalid_zip


np.int64(1)

In [69]:
# Validate that email values contain the required "@" symbol 
invalid_email = df["email"].dropna().apply(
    lambda x: "@" not in str(x)
).sum()

invalid_email

np.int64(8)

In [70]:
'''
The validity analysis identified 1 record with a debt-to-income ratio outside the valid range [0,1], 1 ZIP code containing non-numeric characters, and 8 email addresses missing the required "@" symbol. No negative income values (0), negative interest rates (0), or SSN format violations (0) were detected. 

In total, 10 rule-based validity violations were observed. Although the number is low relative to the dataset size, these records should be explicitly flagged or removed to ensure compliance with defined business constraints.

Option A - Flag invalid values:
df["flag_invalid_dti"] = (df["debt_to_income"] < 0) | (df["debt_to_income"] > 1)
df["flag_invalid_zip"] = ~df["zip_code"].astype(str).str.isdigit()
df["flag_invalid_email"] = ~df["email"].astype(str).str.contains("@", na=False)

Option B - Filter to valid values only:
df_valid = df[
    (df["debt_to_income"] >= 0) & (df["debt_to_income"] <= 1) &
    (df["zip_code"].astype(str).str.isdigit()) &
    (df["email"].astype(str).str.contains("@", na=False))
]
'''

'\nThe validity analysis identified 1 record with a debt-to-income ratio outside the valid range [0,1], 1 ZIP code containing non-numeric characters, and 8 email addresses missing the required "@" symbol. No negative income values (0), negative interest rates (0), or SSN format violations (0) were detected. \n\nIn total, 10 rule-based validity violations were observed. Although the number is low relative to the dataset size, these records should be explicitly flagged or removed to ensure compliance with defined business constraints.\n\nOption A - Flag invalid values:\ndf["flag_invalid_dti"] = (df["debt_to_income"] < 0) | (df["debt_to_income"] > 1)\ndf["flag_invalid_zip"] = ~df["zip_code"].astype(str).str.isdigit()\ndf["flag_invalid_email"] = ~df["email"].astype(str).str.contains("@", na=False)\n\nOption B - Filter to valid values only:\ndf_valid = df[\n    (df["debt_to_income"] >= 0) & (df["debt_to_income"] <= 1) &\n    (df["zip_code"].astype(str).str.isdigit()) &\n    (df["email"].ast

### 4. Accuracy
In this section, accuracy is assessed by evaluating whether recorded values plausibly reflect real-world conditions and logical business relationships. This includes detecting implausible age values, unrealistic spending relative to income, and approvals granted despite zero or missing income, in order to identify records that may be factually incorrect even if they pass basic format and validity checks.

In [71]:
#Accuracy Age
df["date_of_birth"] = pd.to_datetime(df["date_of_birth"], errors="coerce")
df["age"] = (pd.Timestamp("today") - df["date_of_birth"]).dt.days // 365

unrealistic_age = ((df["age"] < 18) | (df["age"] > 100)).sum()
unrealistic_age

np.int64(0)

In [72]:
#Accuracy Spending > Income?
spend_cols = [c for c in df.columns if c.startswith("spending_")]
df["total_spending"] = df[spend_cols].sum(axis=1)

income = pd.to_numeric(df["annual_income"], errors="coerce")

implausible_spending = (df["total_spending"] > income).sum()
implausible_spending

np.int64(1)

In [73]:
#Accuracy Approved loans with Zero or missing income

approved_zero_income = (
    (df["loan_approved"] == True) &
    ((income <= 0) | income.isna())
).sum()

approved_zero_income

np.int64(3)

In [74]:
"""
The accuracy analysis identified 0 unrealistic age values (<18 or >100), 
1 case where total spending exceeds reported annual income, and 
3 approved loans with zero or missing income.

In total, 4 records show plausibility-related accuracy concerns that may reflect 
reporting errors or decision inconsistencies and should be reviewed.

Option A - Flag implausible cases:
df["flag_implausible_spending"] = df["total_spending"] > income
df["flag_approved_zero_income"] = (df["loan_approved"] == True) & ((income <= 0) | income.isna())

Option B - Exclude from modeling:
df_model = df[
    ~(df["total_spending"] > income) &
    ~((df["loan_approved"] == True) & ((income <= 0) | income.isna()))
]
"""

'\nThe accuracy analysis identified 0 unrealistic age values (<18 or >100), \n1 case where total spending exceeds reported annual income, and \n3 approved loans with zero or missing income.\n\nIn total, 4 records show plausibility-related accuracy concerns that may reflect \nreporting errors or decision inconsistencies and should be reviewed.\n\nOption A - Flag implausible cases:\ndf["flag_implausible_spending"] = df["total_spending"] > income\ndf["flag_approved_zero_income"] = (df["loan_approved"] == True) & ((income <= 0) | income.isna())\n\nOption B - Exclude from modeling:\ndf_model = df[\n    ~(df["total_spending"] > income) &\n    ~((df["loan_approved"] == True) & ((income <= 0) | income.isna()))\n]\n'